### ISSUE TITLE
Add String/Line Continuation cops

### ISSUE THREAD

**Issue by pdobb** [rubocop/rubocop#9119]
I believe the below may be more of a category of feature requests, than a single feature request. I will be glad to follow-up and break this out into multiple feature requests, but I am first seeking buy-in and/or direction or advice on if this is even possible to enforce with rubocop and the parser library.

## Is your feature request related to a problem? Please describe.

My team frequently uses line continuation for String concatenation (as mentioned [here, in the Style Guide](https://rubystyle.guide/#heredoc-long-strings)). However:

1. I don't think any cops exist to enforce line continuation style. e.g.

```ruby
# Good (no space before line continuation character).
"part 1 "\  
"part 2"

# Bad (space before line continuation character).
"part 1 " \  
"part 2"
```

```ruby
# Good (conjoining space at end of first line of String).
"part 1 "\  
"part 2"

# Bad (conjoining space at beginning of second line of String).
"part 1"\  
" part 2"

# Bad (conjoining space at both end of first line and beginning of second line of String).
"part 1 "\  
" part 2"
```

```ruby
# Good (second line of String is aligned with start of first line of String).
"part 1 "\  
"part 2"

# Bad (second line of String is misaligned with start of first line of String).
# (Or this could be good and the former could be bad, depending on team preference.)
"part 1 "\  
  "part 2"
```

2. Auto-correction of various cops/offenses can cause manual cleanup work related to line continuation due to no enforcement of the above mentioned styles. e.g.

```ruby
# Given:
# - Layout/FirstMethodArgumentLineBreak is enabled
# - Layout/FirstArgumentIndentation is enabled with:
#   `EnforcedStyle: consistent_relative_to_receiver`
# - Layout/HashAlignment is enabled with
#   `EnforcedHashRocketStyle: [key, table]` and
#   `EnforcedColonStyle: [key, table]`
mail(to: "user@example.com",
     subject: "My super long subject that just keeps going and going and, "\
              "now has to line wrap to fit w/in Layout/LineLength chars.")
```
Then run:
```ruby
> rubocop --auto-correct-all temp.rb
Inspecting 1 file
C

Offenses:

temp.rb:65:6: C: [Corrected] Layout/FirstMethodArgumentLineBreak: Add a line break before the first argument of a multi-line method argument list.
mail(to: "user@example.com",
     ^^^^^^^^^^^^^^^^^^^^^^
temp.rb:66:1: C: [Corrected] Layout/FirstArgumentIndentation: Indent the first argument one step more than mail(.
to: "user@example.com", ...
^^^^^^^^^^^^^^^^^^^^^^^
temp.rb:67:6: C: [Corrected] Layout/HashAlignment: Align the keys of a hash literal if they span more than one line.
     subject: "My super long subject that just keeps going and going and, "\ ...
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 3 offenses detected, 3 offenses corrected
```
Results in a code change of:
```ruby
# Bad (line continuation alignment is now out of whack)
mail(
  to: "user@example.com",
  subject: "My super long subject that just keeps going and going and, "\
                "now has to line wrap to fit w/in Layout/LineLength chars.") # <--- 👀
```
Which then requires manual intervention to go back and find all such issues and fix per the team's style.

3. Cops such as [Layout/BlockAlignment](https://docs.rubocop.org/rubocop/cops_layout.html#layoutblockalignment) are not reasonably usable along with line continuation for String concatenation. e.g.

```ruby
# Given:
# - Layout/IndentationWidth is enabled with default values
# - Layout/BlockAlignment is enabled with 
#   `EnforcedStyleAlignWith: start_of_block`
it "my really long test name that just keeps going and going and "\
   "now has to line wrap to fit w/in Layout/LineLength chars" do
  value(subject.my_method).must_equal("expected value")
end
```
Then run:
```ruby
> rubocop --auto-correct-all temp.rb
Inspecting 1 file
C

Offenses:

temp.rb:72:3: C: [Corrected] Layout/IndentationWidth: Use 2 (not -1) spaces for indentation. (https://rubystyle.guide#spaces-indentation)
  value(subject.my_method).must_equal("expected value")
  ^
temp.rb:73:1: C: [Corrected] Layout/BlockAlignment: end at 73, 0 is not aligned with "now has to line wrap to fit w/in Layout/LineLength chars" do at 71, 3.
end
^^^

1 file inspected, 2 offenses detected, 2 offenses corrected
```
Results in a code change of:
```ruby
# Bad (`end` should be aligned with beginning of `it`)
it "my really long test name that just keeps going and going and "\
   "now has to line wrap to fit w/in Layout/LineLength chars" do
     value(subject.my_method).must_equal("expected value")
   end # <--- 👀
```
Which I imagine is because the line continuation is seen as 2 or multi-lines when really I expect/want it to be treated like a single line. Because just removing the 2nd line of the String in the above example, e.g.
```ruby
it "my really long test name that just keeps going and going and" do
  value(subject.my_method).must_equal("expected value")
end
```
results in no correction, as expected.

## Describe the solution you'd like

I'd love to see a set of cops built for managing all of the above issues with line continuation. I'd be happy to help as much as I can with this, though I am still very green to rubocop's code base and to the parser gem and AST representations in general.

## Describe alternatives you've considered

I've looked for cops dealing with line continuation but have found none.

I haven't yet looked into whether the parser gem's AST might represent (or not) line continuation characters -- which seems key to being able to enforce any of these style choices.

## Additional context

These cops may provide further opportunity for style enforcement of https://rubystyle.guide/#no-trailing-backslash -- to *not* use line continuation in expressions. But this is not an issue that my team, in particular, struggles with.

**Comment by jonas054** [comment]
The requested features are partly implemented now.

### 1

- `Layout/LineContinuationSpacing` (added in 1.31, pending) checks space before backslash. Requires a space by default but is configurable.
- `Layout/LineEndStringConcatenationIndentation` (added in 1.18, pending) checks the indentation of the next line after a line that ends with a string literal and a backslash. Enforces `aligned` by default, but can be set to `indented`.
- `Layout/LineContinuationLeadingSpace` (added in 1.31, pending) checks where the space is in the concatenated strings. Also configurable (on master). Does not specifically report that there is space in both places, though. 

### 2

The example is now corrected to
```ruby
mail(
  to: "user@example.com",
  subject: "My super long subject that just keeps going and going and, " \
           "now has to line wrap to fit w/in Layout/LineLength chars."
)
```

### 3

No change. Still aligns `end` as in the example.

@pdobb Would you close this issue and open a couple of new ones for the remaining problems? I think that would make it easier to fix them.

**Comment by pdobb** [comment]
@jonas054 Looks like great progress, thanks! Unfortunately I'm out of the game now... Given that, I'm happy to just close this issue if it helps? I don't want to push work onto anyone else, but I'm gone as a stakeholder. Thanks again.

**Comment by jonas054** [comment]
I'll close it, then. Given that only minor problems remain. Someone else will report if they're still bothered by these, I'm sure.
